### PR TITLE
[Inductor CUTLASS backend] Step 4: CUDA (template) kernels

### PR DESCRIPTION
**Pull request by ipiszy**
This is the step 4 to add cutlass as an alternative inductor backend.
Full tests can be found from the last PR in the stack.

Feature request: https://github.com/pytorch/pytorch/issues/106991.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #108015
* __->__ #107931
* #107901
* #107847
* #107802

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
...
+        import jinja2
+
+        return jinja2.Environment(
+            undefined=jinja2.StrictUndefined,
+        )
+    except ImportError:
+        return None
+
+
+class ChoiceCaller:
```
Comment (kadeng) [inline torch/_inductor/codegen/common.py:1036]: A small doc comment would be great: What's the purpose of this class / which problem does it solve? Is it supposed to have subclasses?

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
...
+    def hash_key(self):
+        raise NotImplementedError()
+
+    def output_node(self):
+        raise NotImplementedError()
+
+
+class KernelTemplate:
+    """
+    Base class for defining kernel templates.
```
Comment (kadeng) [inline torch/_inductor/codegen/common.py:1062]: Which kind of kernel templates? ( e.g. Triton / C++ / Cutlass / any involving Jinja templates )

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -1029,3 +1019,83 @@ class OptimizationContext:
...
+            if name == fake_out.get_name():
+                return fake_out.get_dtype()
+            return _get_dtype_real(name)
+
+        return get_dtype
+
+    def __init__(self, name: str):
+        self.name = name
+
+    def maybe_append_choice(self, choices, **kwargs):
```
Comment (kadeng) [inline torch/_inductor/codegen/common.py:1086]: What is the "choices" argument here? ( e.g. datatype and intended usage). A small doc comment would help clarify I think.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+from ...ir import Callable, CUDATemplateBuffer, IRNode, Layout, TensorBox
+from ...select_algorithm import ChoiceCaller
+from ...utils import sympy_product
+from ...virtualized import V
+
+from ..common import IndentedBuffer, Kernel, OpOverrides
+from ..cpp import CppPrinter, DTYPE_TO_CPP
+
+
+cexpr = CppPrinter().doprint
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:15]: This might benefit from a type annotation and/or a small comment what it is, so we don't need to follow the link to CppPrinter when reading the code.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+def _normalize_idx(index: int, total_length: int) -> int:
+    return index if index >= 0 else index + total_length
+
+
+class CUDAKernel(Kernel):
+    """
+    Kernels defined by the CUDA language.
+    """
+    overrides = OpOverrides
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:26]: Which consequences does this line have? It's a bit hard to read out of the inductor codegen, how exactly these overrides fields are used.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+        if node is None:
+            return "void"
+        return DTYPE_TO_CPP.get(node.get_layout().dtype)
+
+    def offset(self, node: IRNode) -> str:
+        if node is None:
+            return "0"
+        return str(node.get_layout().offset)
+
+    def ptr(self, node: IRNode, default_node: IRNode = None) -> str:
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:153]: A small doc comment maybe?

File: torch/_inductor/codegen/cuda/cuda_scheduling.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
+from typing import List
+
+from ... import config
+from ...codecache import code_hash, get_path
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/cuda_scheduling.py:4]: It would be better to use absolute imports here, as that eases code navigation tools ( both in IDEs and github code views for example )

File: torch/_inductor/codegen/cuda/cuda_template.py [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+from typing import List
+from unittest.mock import patch
+
+import sympy
+import torch
+
+# import cutlass libs
+import scripts as cutlass_lib
+
+from ...autotune_process import CUDABenchmarkRequest, TensorMeta
```
Comment (kadeng) [inline torch/_inductor/codegen/cuda/cuda_template.py:13]: Same here, absolute imports would be preferable for most tools

File: torch/_inductor/codegen/cuda/cuda_kernel.py (hunk 6 above)
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/cuda_kernel.py]: Added some comments.

File: torch/_inductor/codegen/common.py (hunk 2 above)
Comment (ipiszy) [reply torch/_inductor/codegen/common.py:1070]: Added some comments.

**Review by ipiszy — COMMENTED** [review]
Thanks @kadeng !

File: torch/_inductor/kernel/mm.py [hunk 9]
```diff
@@ -192,15 +179,16 @@ def tuned_addmm(inp, mat1, mat2, *, alpha=1, beta=1, layout=None):
...
-    for config in mm_configs(m, n, k):
-        mm_template.maybe_append_choice(
-            choices,
-            (inp_expanded, mat1, mat2),
-            layout,
-            **mm_options(config, k, layout),
-            prefix_args=1,
-            epilogue_fn=addmm_epilogue(layout.dtype, alpha, beta),
-        )
+    if use_triton_template(layout):
```
Comment (aakhundov) [inline torch/_inductor/kernel/mm.py:182]: Add `m * n != 0` to the condition (now removed from the above)?

File: torch/_inductor/scheduler.py [hunk 10]
```diff
@@ -1644,7 +1644,12 @@ def codegen(self):
 
             if node.is_template():
                 node, *epilogue = node.get_nodes()
-                self.get_backend(device).codegen_template(node, epilogue)
+                if isinstance(node.node, ir.CUDATemplateBuffer):
+                    from .codegen.cuda.cuda_scheduling import CUDAScheduling
+
+                    CUDAScheduling(self).codegen_template(node, epilogue)
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/scheduler.py:1778]: can_fuse() always returns False for CUDATemplateBuffer (in triton.py). It invokes Triton codegen_template() here to codegen only the non-epilogue part.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
...
+        self.named_nodes: Dict[str, IRNode] = {}
+
+    def arg_name(self, node: IRNode) -> Optional[str]:
+        """
+        Returns arg name of a given input or output node.
+        """
+
+        if node is None:
+            return None
+        return {**self.args.input_buffers, **self.args.output_buffers}.get(
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/cuda_kernel.py:55]: OSS linter formats it in this way..

**Review by ipiszy — COMMENTED** [review]
Thanks @aakhundov ! Updated the PR, ptal.

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 12]
```diff
@@ -0,0 +1,309 @@
...
+
+from ..common import IndentedBuffer, Kernel, OpOverrides
+from ..cpp import CppPrinter, DTYPE_TO_CPP
+
+
+cexpr = CppPrinter().doprint
+
+
+def _normalize_idx(index: int, total_length: int) -> int:
+    return index if index >= 0 else index + total_length
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codegen/cuda/cuda_kernel.py]: Actually I don't want to add assertions, as in `stride()` I just return a default value if index is out of bound (useful to get batch_size, batch_stride, etc, from a 2d matrix).

File: test/inductor/test_max_autotune.py [hunk 13]
```diff
@@ -1,323 +0,0 @@
-# Owner(s): ["module: inductor"]
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply test/inductor/test_max_autotune.py]: No, it was a mistake ^^.

File: torch/_inductor/codegen/triton.py [hunk 14]
```diff
@@ -2504,11 +2508,12 @@ def codegen_template(self, template_node, epilogue_nodes):
                 node.codegen(kernel.split_and_set_ranges(node.get_ranges()))
 
         # finalize must be called after adding epilogue above
-        src_code = partial_code.finalize()
+        # TODO: Maybe unify CUDATemplateKernel to also use PartialRender for flexible epilogue fusion.
+        src_code = partial_code if isinstance(partial_code, str) else partial_code.finalize()
```
[the remark this answers is not in the document]
Comment (ipiszy) [reply torch/_inductor/codegen/triton.py]: Both Triton and CUDA execute this code path since CUDAScheduling inherits from TritonScheduling.

File: torch/_inductor/codegen/cuda/cuda_kernel.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (aakhundov) [reply torch/_inductor/codegen/cuda/cuda_kernel.py:17]: Makes sense.

**Comment by aakhundov** [comment]
@ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk.

**Review by aakhundov — CHANGES_REQUESTED** [review]
@ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk. As the 3 PRs below this one in the stack are green and the one above is also red, my guess is that the root cause may be in this one.

**Comment by ipiszy** [comment]
> @ipiszy there are some CI jobs failing, worth checking as not filtered out as flaky / broken trunk. As the 3 PRs below this one in the stack are green and the one above is also red, my guess is that the root cause may be in this one.

Yes on it. There had been a bunch of test failures. I've fixed some and now there is something new. It's an iterative process..

**Review by jansel — APPROVED** [review]

File: torch/_inductor/codegen/cuda/cuda_kernel.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+        """
+        Hook called from template code to generate function def and
+        needed args.
+
+        inputs / outputs: List of input / output IRNodes. Note that IRNode can be None for optional arguments.
+        names_str: Comma separated list of input + output argument names.
+        input_reorder: The actual order of input nodes.
+                       e.g. The template might have input argument defined as [X, W, Bias],
+                       and the actual input passed into this template could be [Bias, X, W].
+                       In this case, the `input_reorder` would be [2, 0, 1].
```
Comment (chenyang78) [inline torch/_inductor/codegen/cuda/cuda_kernel.py:100]: Wondering why we allow different orders for the actual input and the template declaration in the first place? Could we enforce the actual input to have the same order specified by the template?

Comment (aakhundov) [reply torch/_inductor/codegen/cuda/cuda_kernel.py:100]: I assume, for template front-end similarity with the existing Triton templates?

**Review by aakhundov — APPROVED** [review]

**Review by kadeng — APPROVED** [review]
